app/src/main/java/edu/wpi/first/shuffleboard/app/components/Tile.java

```diff
@@ -0,0 +1,109 @@
+package edu.wpi.first.shuffleboard.app.components;
+
+import edu.wpi.first.shuffleboard.api.components.EditableLabel;
+import edu.wpi.first.shuffleboard.api.util.PropertyUtils;
+import edu.wpi.first.shuffleboard.api.util.PseudoClassProperty;
+import edu.wpi.first.shuffleboard.api.widget.Component;
+import edu.wpi.first.shuffleboard.api.widget.Layout;
+import edu.wpi.first.shuffleboard.api.widget.TileSize;
+import edu.wpi.first.shuffleboard.api.widget.Widget;
+
+import org.fxmisc.easybind.EasyBind;
+
+import java.io.IOException;
+
+import javafx.beans.property.BooleanProperty;
+import javafx.beans.property.Property;
+import javafx.beans.property.SimpleObjectProperty;
+import javafx.fxml.FXMLLoader;
+import javafx.scene.layout.BorderPane;
+
+public class Tile<T extends Component> extends BorderPane {
+
+  private final Property<T> content = new SimpleObjectProperty<>(this, "content", null);
+
+  private final Property<TileSize> size = new SimpleObjectProperty<>(this, "size", null);
+  private final BooleanProperty selected = new PseudoClassProperty(this, "selected");
+
+  /**
+   * Creates an empty tile. The content and size must be set with {@link #setContent(T)} and
+   * {@link #setSize(TileSize)}.
+   */
+  protected Tile() {
+    try {
+      FXMLLoader loader = new FXMLLoader(Tile.class.getResource("Tile.fxml"));
+      loader.setRoot(this);
```

Comment:
Is this guaranteed to run in the JavaFX thread?

Also, you are exposing a reference to `this` within the constructor which is bad practice as it can lead to race conditions regarding fields that should be final (and thus non-null) being null when viewed on another thread.

We had this problem with GRIP:
https://github.com/WPIRoboticsProjects/GRIP/issues/155


[Guava Google Groups Discussion with Dev](https://groups.google.com/forum/#!topic/guava-discuss/4ZJhhpmMvKQ)

> ## Don't publish the "this" reference during construction
> 
> One of the mistakes that can introduce a data race into your class is to expose the this reference to another thread before the constructor has completed. Sometimes the reference is explicit, such as directly storing this in a static field or collection, but other times it can be implicit, such as when you publish a reference to an instance of a non-static inner class in a constructor. Constructors are not ordinary methods -- they have special semantics for initialization safety. An object is assumed to be in a predictable, consistent state after the constructor has completed, and publishing a reference to an incompletely constructed object is dangerous. Listing 2 shows an example of introducing this sort of race condition into a constructor. It may look harmless, but it contains the seeds of serious concurrency problems.
> -[Brian Goetz](http://www.ibm.com/developerworks/library/j-jtp0618/)

Reply:
this is the correct and documented way to construct JavaFX custom components. see https://docs.oracle.com/javase/8/javafx/fxml-tutorial/custom_control.htm